scripts/apply_style.py

```diff
@@ -104,233 +110,277 @@ def output(text):
 
-    output('Checking header include guards ...')
-    fix_missing_include_guards()
+def files_of_interest(commit):
+    """Returns list of absolute paths to files"""
+    global TOTAL_FILES_TO_PROCESS
+
+    output('Composing list of files ...')
+
+    ret = []
+
+    if commit is None:
+        for root, _, files in walk_source_directories(PROJECT_ROOT):
+            for file_name in files:
+                absolute_path = abspath(join(root, file_name))
+                ret.append(absolute_path)
+    else:
+        ret = get_changed_paths_from_git(commit)
+
+    # Filter out binary files and symbolic links
+    ret = [abs_path for abs_path in ret
+           if isfile(abs_path) and not is_checked_in_binary_file(abs_path)]
+
+    TOTAL_FILES_TO_PROCESS = len(ret)
+
+    return ret
+
+
+def print_diff_and_fail():
+    diff = subprocess.check_output(
+        ['git', 'diff'], cwd=PROJECT_ROOT).decode().strip()
+    if diff:
+        output('*' * 80)
+        output(diff)
+        output('*' * 80)
+        sys.exit(1)
+
+
+def parse_commandline():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        '-c',
+        '--commit',
+        nargs=1,
+        default=None,
+        help='scan and fix only files that changed between Git\'s HEAD and the '
+             'given commit or ref. If not set, process all files in the project')
+    parser.add_argument(
+        '-d',
+        '--diff',
+        action='store_true',
+        help='if deviations from the style are found, print the diff and exit '
+             'with error. Useful for automated checks')
+    parser.add_argument(
+        '-j',
+        '--jobs',
+        type=int,
+        default=min(multiprocessing.cpu_count(), 7),
+        help='the maximum number of parallel jobs')
+
+    args = parser.parse_args()
+
+    return args.commit if args.commit is None else args.commit[0], \
+        args.diff, \
+        args.jobs
+
+
+TRANSFORMATIONS = [
+    fix_cmake_version_requirements,
+    fix_license_header,
+    fix_missing_include_guards,
+    fix_terminal_newlines,
+    external_formatter(**SUPPORTED_LANGUAGES['cpp']),
+    external_formatter(**SUPPORTED_LANGUAGES['cmake']),
+    format_python
+]
+
+
+def apply_transformations_to_file(path_to_file):
+    try:
+        with open(path_to_file, 'r+', encoding='utf-8') as f:
+            original_text = f.read()
+            text = original_text
+
+            for transformation in TRANSFORMATIONS:
+                text = transformation(text, path_to_file)
+
+            if text != original_text:
+                f.seek(0)
+                f.truncate(0)
+                f.write(text)
+
+        increment_file_count()
+    except:
+        output(traceback.format_exc())
+
+
+def process_in_parallel(files_to_process, jobs):
+    output('Processing {} file(s) with {} job(s) ...'
+           .format(TOTAL_FILES_TO_PROCESS, jobs))
+    with ThreadPoolExecutor(max_workers=jobs) as pool:
```

Comment:
I think you will surprised but I don't think this is going to do what want because of python's GIL. (in the case of the calling external tools it is fine). It would be probably worth switching to `ProcessPoolExecutor`

Reply:
That's done. The performance increase is quite modest, as profiling indicates that about 95% of the time is spent calling out to the external formatters anyway. I had trouble finding a way of synchronising the progress indicators when using `ProcessPoolExecutor`, so I switched to using lower level `multiprocessing` utilities.